template/src/utils/api.ts

```diff
@@ -0,0 +1,87 @@
+export enum RemoteDataStates {
+  NOT_REQUESTED = 'NOT_REQUESTED',
+  LOADING = 'LOADING',
+  REFRESHING = 'REFRESHING',
+  FAILURE = 'FAILURE',
+  SUCCESS = 'SUCCESS',
+}
+
+type NotRequestedType = {
+  state: RemoteDataStates.NOT_REQUESTED
+}
+
+type LoadingType = {
+  state: RemoteDataStates.LOADING
+}
+
+type RefreshingType<T> = {
+  state: RemoteDataStates.REFRESHING
+  data: T
+}
+
+type FailureType<E, T> = {
+  state: RemoteDataStates.FAILURE
+  error: E
+  data?: T
+}
+
+type SuccessType<T> = {
+  state: RemoteDataStates.SUCCESS
+  data: T
+}
+
+// Constructors
+
+export const NotRequested: NotRequestedType = {
+  state: RemoteDataStates.NOT_REQUESTED,
+}
+
+export const Loading: LoadingType = { state: RemoteDataStates.LOADING }
+
+export const Refreshing = <T>(data: T): RefreshingType<T> => ({
+  state: RemoteDataStates.REFRESHING,
+  data,
+})
+
+export const Pending = <T, E>(remoteData: RemoteData<T, E>) => {
+  return hasData(remoteData) ? Refreshing(remoteData.data) : Loading
+}
+
+export const Failure = <E, T>(error: E, data?: T): FailureType<E, T> => ({
+  state: RemoteDataStates.FAILURE,
+  error,
+  ...(data && { data }),
+})
+
+export const Success = <T>(data: T): SuccessType<T> => ({
+  state: RemoteDataStates.SUCCESS,
+  data,
+})
+
+export type RemoteData<T, E> =
+  | NotRequestedType
+  | LoadingType
+  | RefreshingType<T>
+  | FailureType<E, T>
+  | SuccessType<T>
+
+export const isNotRequested = <T, E>(
+  remoteData: RemoteData<T, E>,
+): remoteData is NotRequestedType => remoteData.state === RemoteDataStates.NOT_REQUESTED
+
+export const isLoading = <T, E>(remoteData: RemoteData<T, E>): remoteData is LoadingType =>
+  remoteData.state === RemoteDataStates.LOADING
+
+export const isRefreshing = <T, E>(remoteData: RemoteData<T, E>): remoteData is RefreshingType<T> =>
+  remoteData.state === RemoteDataStates.REFRESHING
+
+export const isFailure = <T, E>(remoteData: RemoteData<T, E>): remoteData is FailureType<E, T> =>
+  remoteData.state === RemoteDataStates.FAILURE
+
+export const isSuccess = <T, E>(remoteData: RemoteData<T, E>): remoteData is SuccessType<T> =>
+  remoteData.state === RemoteDataStates.SUCCESS
+
+export const hasData = <T, E>(
```

Comment:
what would be the usecase of this check?

Reply:
1. lets say that we have object of type `RemoteData`. Until you do this check on it, typescript won't be happy if you try to access `data` property on it, saying that there is no `data` field on `NotRequested` / `LoadingType`, cause these types are part of the `RemoteData` union too.
2. Lets say you have `RemoteData` object and you want to display whatever it is and you don't care whether its state is a `success` / `refreshing` or `failure`.

Keep in mind its not a new code, but also happy to know if there is something in it that bothers you. It's not the first time when we want to fix it for good, but it still happen to come with some unresolved drawbacks.

Reply:
looks good, it wasnt bothering me, rather wanted to refamiliarize myself with the template :)

Reply:
thinking about it now makes me wonder whether I shouldnt *blazenly* use this stuff in our project ;D

Reply:
I really like this concept, but on the other hand there are still some problems with it, that I didn't find solution for.
especially second one here -> https://nozbe.app/teams/Xi3SI50NCpPkVxNz/task_id/3sAjCbeo8gJzHjcA